### DISCUSSION TITLE
Remove a indexed file from “Languages” on first page

### DISCUSSION THREAD

**Discussion by pownas** [github-linguist/linguist#5175] [Classification] [answered]
Hi! 
I found now that you have a Discussion in this repo. 
Maybe you can help me answer my question on StackOverflow? 
https://stackoverflow.com/questions/65905812/github-remove-a-indexed-file-from-languages-on-first-page

The question below: 

How can I remove this indexed HTML page, that are a documentation to one of the external librarys I use on my GitHub blob? 

I have tried alot of diffrent commands, but don't find a way to remove this file from the [GitHub Linguist][1] indexer...

Here are the "Languages" that are indexed on the startpage: 
![[image] Languages on the startpage](https://i.stack.imgur.com/MpmXI.png)

And the file that I want to exclude: 
```
TestProject/wwwroot/lib/bootstrap-icons/docs/index.html
```
![[image] HTML file that needs to be excluded](https://i.stack.imgur.com/GYp4c.png)

Code that I've tried to get it removed via ```.attributes```-file in root-folder (the vendored, works... But not getting rid of this HTML-file... from the GitHub-Languages) : 
```
### vendored:
TestProject/wwwroot/lib/* linguist-vendored

### documentations:
TestProject/wwwroot/lib/bootstrap-icons/* linguist-documentation
```
and tried: 
```
TestProject/wwwroot/lib/bootstrap-icons/* -linguist-documentation
```
and this: 
```
TestProject/wwwroot/lib/bootstrap-icons/docs/* linguist-documentation
```
and this: 
```
TestProject/wwwroot/lib/bootstrap-icons/docs/* -linguist-documentation
```
and this:
```
TestProject/wwwroot/lib/* linguist-documentation
```
and this: 
```
TestProject/wwwroot/lib/* -linguist-documentation
```

But I can't figure it out how to remove this file: 
```
TestProject/wwwroot/lib/bootstrap-icons/docs/index.html
```

Please help me with the correct syntax to remove this file from being indexed as a Language in my GitHub repository, main branch. 🙂
Or please add in the documentation the syntax or a example .yml/.attributes-file for how to override and remove one specific file or folder from the language. 

  [1]: https://github.com/github/linguist#overrides

**Comment by pownas** [comment]
Current .attributes file (that still shows the HTML file in the "docs"-folder: 

![image](https://user-images.githubusercontent.com/47676607/106125389-67207d00-615c-11eb-8d46-8beadcecc9e1.png)


And the file that I need removed: 

![image](https://user-images.githubusercontent.com/47676607/106125691-bb2b6180-615c-11eb-9c66-81ba258df97f.png)

**Comment by lildude** [comment, marked as answer]
I'll answer in both places with the same response:

You've got the right idea and the right Linguist overrides (either will do the trick). The problem is your path matching isn't quite right.

From the [`.gitattributes` docs](https://www.git-scm.com/docs/gitattributes)

> The rules by which the pattern matches paths are the same as in `.gitignore` files (see gitignore[5]), with a few exceptions:
> [...]

If we look in the [`.gitignore` docs](https://www.git-scm.com/docs/gitignore) (emphasis is mine):

> An asterisk "`*`" matches anything **except a slash**. The character "`?`" matches any one character except "`/`". The range notation, e.g. `[a-zA-Z]`, can be used to match one of the characters in a range. See fnmatch(3) and the FNM_PATHNAME flag for a more detailed description.
> 
> Two consecutive asterisks ("`**`") in patterns matched against full pathname may have special meaning:
> [...]
> - A trailing "`/**`" matches everything inside. For example, "`abc/**`" matches all files inside directory "`abc`", relative to the location of the `.gitignore` file, with infinite depth.

The files you're trying to ignore are in sub-directories of the paths you've specified so you need to either:

- use `TestProject/wwwroot/lib/** linguist-vendored` to recurse, or
- use `TestProject/wwwroot/lib/bootstrap-icons/docs/* linguist-vendored` to limit to this directory.

We can demonstrate this without even using Linguist thanks to `git check-attr`:

```console
$ # Create a repo with just the one file
$ git init -q Test-Project
$ cd Test-Project
$ mkdir -p TestProject/wwwroot/lib/bootstrap-icons/docs/
$ echo "<html>" > TestProject/wwwroot/lib/bootstrap-icons/docs/index.html
$ git add -A
$ git commit -m 'Add file'
[main (root-commit) bed71b5] Add file
 1 file changed, 1 insertion(+)
 create mode 100644 TestProject/wwwroot/lib/bootstrap-icons/docs/index.html
$
$ # Add your initial override
$ git add -A && git commit -m 'attribs'
[main 7d0a0cf] attribs
 1 file changed, 1 insertion(+)
 create mode 100644 .gitattributes
$
$ # Check the attributes
$ git check-attr linguist-vendored TestProject/wwwroot/lib/bootstrap-icons/docs/index.html
TestProject/wwwroot/lib/bootstrap-icons/docs/index.html: linguist-vendored: unspecified
$ # So it doesn't have any effect.
$ # Now lets recurse
$ echo "TestProject/wwwroot/lib/** linguist-vendored" > .gitattributes
$ git add -A && git commit -m 'attribs'
[main 9007c34] attribs
 1 file changed, 1 insertion(+), 1 deletion(-)
$ git check-attr linguist-vendored TestProject/wwwroot/lib/bootstrap-icons/docs/index.html
TestProject/wwwroot/lib/bootstrap-icons/docs/index.html: linguist-vendored: set
$ # Woohoo!!! It's work. 
$ # Lets be specific to the docs dir
$ echo "TestProject/wwwroot/lib/bootstrap-icons/docs/* linguist-vendored" > .gitattributes
$ git add -A && git commit -m 'attribs'
[main a46f416] attribs
 1 file changed, 1 insertion(+), 1 deletion(-)
$ git check-attr linguist-vendored TestProject/wwwroot/lib/bootstrap-icons/docs/index.html
TestProject/wwwroot/lib/bootstrap-icons/docs/index.html: linguist-vendored: set
$ # Woohoo!!! It's worked too
```

**Reply by pownas** [reply]
Thanks for your really good instructions!
But I still can't get it to work. 😞 


Please checkout this repo where the HTML-language still are shown on main:
https://github.com/pownas/Test-Project

**Reply by lildude** [reply]
You _have_ got it to work, and your override is definitely excluding the HTML files you wanted, though this isn't clear because of the way Linguist groups HTML-like languages under HTML and the way the search works when you click the language in the sidebar.

The reason you're still seeing HTML is because of all those `.cshtml` files in your repo (the language has been detected as "HTML+Razor") as can be see with [this search](https://github.com/pownas/Test-Project/search?l=HTML%2BRazor). Linguist groups these under "HTML" when considered for the statistics as shown in the sidebar.

The search functionality that powers the link behind the language name doesn't know about the grouping so it searches for the group name as if it were the actual language name.  "HTML+Razor" with the link I've used above is what would be shown in the sidebar if Linguist didn't group all the HTML-like languages under "HTML".

**Comment by pownas** [comment]
One ugly workaround to get the file to be not shown, seems to be: 

Move the file from: 
{root}\TestProject\wwwroot\lib\bootstrap-icons\docs\index.html
to:
{root}}\docs\index.html

and maybe rename it: 
{root}}\docs\bootstrap-icons.html

**Reply by lildude** [reply]
As I mentioned above, your current override is working and is already marking those files as vendored (you don't need the documentation override) so no workaround is needed.

**Reply by pownas** [reply]
Yes, it marks the files as vendored via: 
```TestProject/wwwroot/lib/** linguist-vendored```


But the ```*.html```-file aint seem to be ignored: 
![image](https://user-images.githubusercontent.com/47676607/106159619-4a4c6f80-6185-11eb-806d-0fa8af8ca73a.png)



If I remove the line: 
```TestProject/wwwroot/lib/** linguist-vendored```
All the vendored ```*.css```/```*.js```-files will show up also: 
![image](https://user-images.githubusercontent.com/47676607/106160037-b4651480-6185-11eb-8795-52e498aee2b1.png)

**Reply by lildude** [reply]
They are. As I mentioned at https://github.com/github/linguist/discussions/5175#discussioncomment-317951 that is for all the CSHTML files in your repo.

Here's the analysis Linguist performs:

```console
$ git clone -q git@github.com:pownas/Test-Project.git
$ cd Test-Project/
$ github-linguist . --breakdown
54.53%  C#
42.76%  HTML
1.75%   SCSS
0.97%   JavaScript

C#:
TestProject/Areas/Identity/IdentityHostingStartup.cs
TestProject/Areas/Identity/Pages/Account/AccessDenied.cshtml.cs
[... truncated for brevity ...]
TestProject/Startup.cs
TestProject/Views/Shared/UploadAdminNavPages.cs

HTML:
TestProject/Areas/Identity/Pages/Account/AccessDenied.cshtml
TestProject/Areas/Identity/Pages/Account/ConfirmEmail.cshtml
TestProject/Areas/Identity/Pages/Account/ConfirmEmailChange.cshtml
TestProject/Areas/Identity/Pages/Account/ExternalLogin.cshtml
TestProject/Areas/Identity/Pages/Account/ForgotPassword.cshtml
TestProject/Areas/Identity/Pages/Account/ForgotPasswordConfirmation.cshtml
TestProject/Areas/Identity/Pages/Account/Lockout.cshtml
TestProject/Areas/Identity/Pages/Account/Login.cshtml
TestProject/Areas/Identity/Pages/Account/LoginWith2fa.cshtml
TestProject/Areas/Identity/Pages/Account/LoginWithRecoveryCode.cshtml
TestProject/Areas/Identity/Pages/Account/Logout.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/ChangePassword.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/Disable2fa.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/DownloadPersonalData.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/Email.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/EnableAuthenticator.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/ExternalLogins.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/GenerateRecoveryCodes.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/Index.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/PersonalData.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/ResetAuthenticator.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/SetPassword.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/ShowRecoveryCodes.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/TwoFactorAuthentication.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/_Layout.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/_ManageNav.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/_StatusMessage.cshtml
TestProject/Areas/Identity/Pages/Account/Manage/_ViewImports.cshtml
TestProject/Areas/Identity/Pages/Account/Register.cshtml
TestProject/Areas/Identity/Pages/Account/RegisterConfirmation.cshtml
TestProject/Areas/Identity/Pages/Account/ResendEmailConfirmation.cshtml
TestProject/Areas/Identity/Pages/Account/ResetPassword.cshtml
TestProject/Areas/Identity/Pages/Account/ResetPasswordConfirmation.cshtml
TestProject/Areas/Identity/Pages/Account/_StatusMessage.cshtml
TestProject/Areas/Identity/Pages/Account/_ViewImports.cshtml
TestProject/Areas/Identity/Pages/Error.cshtml
TestProject/Areas/Identity/Pages/_ValidationScriptsPartial.cshtml
TestProject/Areas/Identity/Pages/_ViewImports.cshtml
TestProject/Areas/Identity/Pages/_ViewStart.cshtml
TestProject/Views/Bilder/Index.cshtml
TestProject/Views/Bilder/Sök.cshtml
TestProject/Views/Bilder/_Gallery.cshtml
TestProject/Views/Home/Index.cshtml
TestProject/Views/Home/Privacy.cshtml
TestProject/Views/Info/Copyright.cshtml
TestProject/Views/Info/Gästbok.cshtml
TestProject/Views/Info/Index.cshtml
TestProject/Views/Info/Kontakta.cshtml
TestProject/Views/Info/Köp_av_bilder.cshtml
TestProject/Views/Info/Om_mig.cshtml
TestProject/Views/Info/Sidkarta.cshtml
TestProject/Views/Info/_ContactForm.cshtml
TestProject/Views/Senast/Index.cshtml
TestProject/Views/Shared/Error.cshtml
TestProject/Views/Shared/_Footer.cshtml
TestProject/Views/Shared/_Layout.cshtml
TestProject/Views/Shared/_NavBar.cshtml
TestProject/Views/Shared/_NavLoginPartial.cshtml
TestProject/Views/Shared/_NavUploadAdmin.cshtml
TestProject/Views/Shared/_ValidationScriptsPartial.cshtml
TestProject/Views/_ViewImports.cshtml
TestProject/Views/_ViewStart.cshtml

SCSS:
TestProject/wwwroot/css/site.scss

JavaScript:
TestProject/wwwroot/js/ContactForm.js
TestProject/wwwroot/js/gLightBoxOptions.js
TestProject/wwwroot/js/site.js

$
```

As you can see, _none_ of the files under the HTML category are those under `TestProject/wwwroot/lib/` because you've marked those to be ignored. The files under HTML are all the CSHTML files.

**Reply by lildude** [reply]
This is where the `.cshtml` files are classified in Linguist:

https://github.com/github/linguist/blob/299318747a903a90581f46e637f7d3be387ff844/lib/linguist/languages.yml#L2141-L2153

Note the `group` line; that's what's responsible for those files being classified as HTML and for the "HTML" in the sidebar.

**Reply by pownas** [reply]
Oh good, I'm stupid here... 

I see now.... Thanks alot for your patience! 🥇 

The trouble is that the link from the first page, says HTML and points to: 
https://github.com/pownas/Test-Project/search?l=html

When it groups all the .cshtml files under HTML also and the link on First-page aint saying "HTML+Razor" or showing to: 
https://github.com/pownas/Test-Project/search?l=HTML%2BRazor


Thanks alot man! 👍🏻
